### DISCUSSION TITLE
Can Dafoam be renewed

### DISCUSSION THREAD

**Discussion by 15175829310** [mdolab/dafoam#697] [General]
Can Dafoam be renewed if it stops midway

**Comment by friedenhe** [comment]
Yes, you need to prescribe initial design variables in runScript.py. By default, the initial design variables are zeros (shape optimization) or ones (field inversions). You can prescribe the latest design variables from a half-way-finished optimization as the initial design variables to start a new optimization. This way, it is kind of like (not the same) a re-started optimizaiton.

**Reply by zhangzhen117** [reply]
Hi Prof. He, I tried to set the initial value of DVs for the field inversion of fvSource (vector). I read the openfoam field, transfered it into a numpy array, and set it as the initial value. `self.dvs.add_output("fvSource", val=fvSourceValue.reshape(-1))` I can succeed in serial, but I failed in parallel. The field of fvSource in parallel is not the same as what I expect, maybe due to the MPI. How can I fix it? Thank you!

**Comment by friedenhe** [comment]
@zhangzhen117. I suspect the error you had was related to variable ordering. Check this post: https://github.com/mdolab/dafoam/discussions/102

**Reply by zhangzhen117** [reply]
Yes, it is related. The solution there is to read the parallel result, but what I want to achieve is to write an initial field that the parallel function can read. Can I initialize the dv field from the OpenFOAM files? I can let the code read the fvSource from time 0, but it will be overwritten by the value in DV. Can I somehow avoid it?

**Comment by friedenhe** [comment]
If you need to do that, you should read the fields from OpenFOAM files in parallel, i.e., read them from the processor* folders. Then, gather the fields from all the processors and combine them as one field and set them to the DV. BTW: why would you need to read initial fvSource fields?

**Comment by zhangzhen117** [comment]
Thank you. How can I combine fields from all processors? In [#102](https://github.com/mdolab/dafoam/discussions/102) you said it is not simply adding processor0/alphaPorosity, processor1/alphaPorosity, processor2/alphaPorosity, processor3/alphaPorosity. 
Besides, I noticed that the function `void DAInputField::run(const scalarList& input)` assigns values from DAinput to the field, which is the inverse of what I want to do. It seems I can use this part to read DAinput from openfoam.
As for reading initial dvs, because I want to simultaneously match force and velocity. It would be easier to start from some good initial guess of the force(noisy measurement, for instance), so I want to read it.

**Reply by zhangzhen117** [reply]
I tested the scalar case, and the simple combination works. By reading fields from all processors and combining them following the order of processors, I can recover the initial scalar field in parallel. I will also test the vector fields.

**Reply by zhangzhen117** [reply]
I also tested the vector case, and the simple combination still works. The order is Vx Vy Vz components for one point, and following the point order within one processor, and following the order of processors. Thank you!

**Comment by zhangzhen117** [comment]
BTW, reading parameters of an NN regression model in parallel is no problem. So I can always read NN parameters. To read the fvSource field is for validating, because I think field inversion is more accurate than training a regression model, considering the limited input features and the limited expression ability of the NN in this case.

**Comment by friedenhe** [comment]
It is good to know that combing all the fields worked.

**Reply by zhangzhen117** [reply]
Thank you for the information!
